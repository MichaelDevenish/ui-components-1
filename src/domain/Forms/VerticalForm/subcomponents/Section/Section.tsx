import React from 'react'

import { Props, Variables } from '../../../../../common'
import { Margin } from '../../../../Spacers/Margin'
import { Text } from '../../../../Typographies/Text'

interface ISectionProps {
  /** Name to display */
  name: string
  /** Display an additional component to the right of the name */
  rightComponent?: JSX.Element | JSX.Element[] | string
  /** The data-component-context */
  componentContext?: string
}

const Section: React.FC<ISectionProps> = ({ name, componentContext, rightComponent, children }) => {
  return (
    <div
      data-component-type={Props.ComponentType.VerticalFormSection}
      data-component-context={componentContext}
    >
      <Margin margins={{ bottom: Variables.Spacing.sSmall }}>
        <Text
          type={Props.TypographyType.Heading}
          margins={rightComponent ? { right: Variables.Spacing.sXSmall } : undefined}
        >
          {name}
        </Text>
        {rightComponent}
      </Margin>
      {children}
    </div>
  )
}

export {
  Section,
  ISectionProps
}
