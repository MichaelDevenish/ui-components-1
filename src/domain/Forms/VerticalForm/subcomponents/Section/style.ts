import styled, { css } from 'styled-components'

import { Variables } from '../../../../../common'

const StyledSectionWrapper = styled.div`
  margin-bottom: ${Variables.Spacing.sXLarge}px;
`
export {
  StyledSectionWrapper
}
