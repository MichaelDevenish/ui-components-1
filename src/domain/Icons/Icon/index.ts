export { Icon, IIconProps } from './Icon'
export {
  IconType,
  FontAwesomeIconType,
  FontAwesomeIconTypeNoPrefix,
  IntelliIconType,
  IntelliIconTypeNoPrefix
} from './types'
