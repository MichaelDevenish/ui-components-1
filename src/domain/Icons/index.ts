export * from './FontAwesomeIcon'
export * from './IntelliIcon'
export * from './Icon'
