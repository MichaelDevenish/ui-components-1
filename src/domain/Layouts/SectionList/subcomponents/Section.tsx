import React from 'react'

import {Props} from '../../../../common'
import {StyledSection} from '../style'

export interface ISectionProps {
  /** The data-component-context */
  componentContext?: string
}

class Section extends React.PureComponent<ISectionProps> {
  public render (): JSX.Element {
    const {
      children,
      componentContext
    } = this.props

    return (
      <StyledSection
        data-component-type={Props.ComponentType.Section}
        data-component-context={componentContext}
      >
        {children}
      </StyledSection>
    )
  }
}

export {
  Section
}
