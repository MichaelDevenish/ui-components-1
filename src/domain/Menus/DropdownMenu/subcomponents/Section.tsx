import React from 'react'
import { Anchor } from '../../../Internals'
import { StyledSection } from './style'

type SectionType =
  'default'
  | 'alert'
  | 'stripAlert'
  | 'stripSuccess'
  | 'stripWarning'
  | 'stripPrimary'
  | 'stripSecondary'
  | 'stripNeutral'

interface ISectionProps {
  /** What text to show in a section. Will only be used if no component is provided. */
  text?: string,
  /** A component that is shown to the left of the text */
  leftComponent?: JSX.Element,
  /** A component that is shown to the right of the text */
  rightComponent?: JSX.Element,
  /** Determines how to style this section, allowing for different styles for alerts and so on */
  sectionType?: SectionType,
  /** A link that will be navigated to on click */
  href?: string,
  /** Event handler when the section is clicked */
  onClick?: (event: React.SyntheticEvent<HTMLLIElement>) => void
  /**
   * Should this section close the menu when clicked?
   * Defaults to closing when onClick or href is provided; can overwrite with this prop
   */
  closeDropdownBehaviour?: 'always' | 'never' | 'whenActionProvided',
  /** An override component that will render instead of the inbuilt components */
  component?: JSX.Element,
  /** Any extra props to pass to the component */
  componentProps?: any,
  /** Internal prop used for closing the section */
  __closeMenuCallback?: (event: React.SyntheticEvent<HTMLLIElement>, section: ISectionProps) => void
}

class Section extends React.PureComponent<ISectionProps, never> {
  private get component () {
    const {
      onClick,
      href
    } = this.props

    if (href) {
      return Anchor
    }

    if (onClick) {
      return 'button'
    }

    return 'span'
  }

  private get styleProps () {
    const {
      onClick,
      href,
      sectionType
    } = this.props

    return {
      clickable: !!href || !!onClick,
      sectionType: sectionType || 'default',
      onClick: this.handleCloseMenuClick,
      role: 'menuitem'
    }
  }
  public static defaultProps: Partial<ISectionProps> = {
    sectionType: 'default',
    closeDropdownBehaviour: 'whenActionProvided'
  }

  public render (): JSX.Element {
    const {
      component,
      componentProps,
      leftComponent,
      rightComponent,
      text,
      onClick,
      href
    } = this.props

    if (component) {
      return (
        <StyledSection {...this.styleProps}>
          {component}
        </StyledSection>
      )
    }

    const Component = this.component
    return (
      <StyledSection{...this.styleProps}>
        <Component
          {...componentProps}
          onClick={onClick}
          href={href}
        >
          {leftComponent && <span className='left-component'>{leftComponent}</span>}
          {text}
          {rightComponent && <span className='right-component'>{rightComponent}</span>}
        </Component>
      </StyledSection>
    )
  }

  private handleCloseMenuClick = (event: React.SyntheticEvent<HTMLLIElement>) => {
    const {
      href,
      onClick,
      closeDropdownBehaviour,
      __closeMenuCallback
    } = this.props

    if (closeDropdownBehaviour === 'always') {
      __closeMenuCallback!(event, this.props)
    }

    if (closeDropdownBehaviour === 'whenActionProvided' && (!!href || !!onClick)) {
      __closeMenuCallback!(event, this.props)
    }
  }
}

export {
  Section,
  ISectionProps,
  SectionType
}
