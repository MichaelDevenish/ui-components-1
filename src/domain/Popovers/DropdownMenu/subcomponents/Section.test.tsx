import { mount, shallow } from 'enzyme'
import React from 'react'

import { FontAwesomeIcon } from '../../../Icons'
import { Section } from './Section'

describe('<Section />', () => {
  describe('Default sections', () => {
    const wrapper = shallow(
      <span>
        <Section
          text='Item 1'
        />
        <Section
          text='Item 2'
          onClick={jest.fn()}
        />
        <Section
          text='Item 3'
          href='https://intellihr.com.au'
          componentProps={{
            style: { fontWeight: 600 }
          }}
        />
        <Section
          component={<hr />}
        />
        <Section
          text='Item 5'
          leftComponent={<FontAwesomeIcon type='regular' icon='hand-point-right' />}
          rightComponent={<FontAwesomeIcon type='regular' icon='hand-point-left' />}
        />
      </span>
    )

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Left and Right sections', () => {
    const wrapper = mount(
      <Section
        text='surrounded'
        leftComponent={<FontAwesomeIcon type='regular' icon='hand-point-right' />}
        rightComponent={<FontAwesomeIcon type='regular' icon='hand-point-left' />}
      />
    )

    it('should render the left and right components', () => {
      const parent = wrapper
        .find({ text: 'surrounded' })

      expect(
        parent
          .find('.fa-hand-point-right')
          .exists()
      ).toBeTruthy()

      expect(
        parent
          .find('.fa-hand-point-left')
          .exists()
      ).toBeTruthy()
    })
  })

  describe('Custom Components', () => {
    const wrapper = mount(
      <Section
        component={<div>asdf</div>}
      />
    )

    it('should contain a custom element', () => {
      expect(
        wrapper
          .find('li')
          .find('div')
          .filterWhere((element) => element.text() === 'asdf')
          .exists()
      ).toBeTruthy()
    })
  })
})
