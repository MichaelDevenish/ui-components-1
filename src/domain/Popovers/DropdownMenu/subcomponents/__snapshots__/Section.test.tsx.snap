// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Section /> Default sections should match the snapshot 1`] = `
<span>
  <Section
    closeDropdownBehaviour="whenActionProvided"
    sectionType="default"
    text="Item 1"
  />
  <Section
    closeDropdownBehaviour="whenActionProvided"
    onClick={[MockFunction]}
    sectionType="default"
    text="Item 2"
  />
  <Section
    closeDropdownBehaviour="whenActionProvided"
    componentProps={
      Object {
        "style": Object {
          "fontWeight": 600,
        },
      }
    }
    href="https://intellihr.com.au"
    sectionType="default"
    text="Item 3"
  />
  <Section
    closeDropdownBehaviour="whenActionProvided"
    component={<hr />}
    sectionType="default"
  />
  <Section
    closeDropdownBehaviour="whenActionProvided"
    leftComponent={
      <FontAwesomeIcon
        type="hand-o-right"
      />
    }
    rightComponent={
      <FontAwesomeIcon
        type="hand-o-left"
      />
    }
    sectionType="default"
    text="Item 5"
  />
</span>
`;
